teps/draft-20200716-task-results-in-final-tasks.md

```diff
@@ -0,0 +1,279 @@
+---
+title: task-results-in-final-tasks
+authors:
+  - "@pritidesai"
+  - "@bobcatfish"
+creation-date: 2020-07-16
+last-updated: 2020-07-16
+status: proposed
+---
+
+# TEP-NNNN: Task Results in Final Tasks
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+- [Design Details](#design-details)
+- [Advantages](#advantages)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+A `Task` in a `Pipeline` can emit `Results` which can be consumed by any other `Tasks` within the same
+`Pipeline`. `Result` of a `Task` in a `Pipeline` is received through the input parameter using variable
+substitution such as `$(tasks.<pipeline-task-name>.results.<result-name>)`.
+
+Today, final `Tasks` can not receive `Results` of any other `Tasks` in a `Pipeline`. This feature is
+limited to non-finally `Tasks` of a `Pipeline`.
+
+This proposal is enabling final `Tasks` to receive `Results` of any `Tasks` within the same `Pipeline`.
+
+## Motivation
+
+A `Pipeline` author wants to clean up resources that is done by the final `Tasks` in the `Pipeline`.
+The resources are acquired by the setup `Task` which has the names of resources. The clean up `Tasks`
+must clean up only those resources.
+
+Also, cleanup must be done only if the setup was successful but regardless of the failure of the rest of
+the `Tasks` in the `Pipeline`.
+
+```none
+setup
+  |
+  | e.g. Project Name
+  v
+cleanup
+```
+
+### Goals
+
+One goal of this proposal is to allow final `Tasks` to receive `Results` from non-finally `Tasks`.
+
+Second goal of this proposal is to determine what happens to the final `Task` when referenced
+task results are not initialized because either dependent task failed or dependent task skipped
+due to its own condition or if its parent was skipped.
+
+### Non-Goals
+
+This proposal is not enabling final `Task` to receive `Results` from other final `Tasks`.
+
+This proposal is also not enabling final `Tasks` to produce `Task` `Results`.
+
+## Proposal
+
+Allow a finally `Task` to get a value from a non-finally `Task` via a `Task` `Result`.
+If the `Task` producing the value fails, the finally `Task` that requires that value will be marked as failed.
+That final `Task` will be `attempted` i.e. evaluated by the `Pipeline Controller` for execution,
+but the validation will fail before the `TaskRun` is created since the required `Result` is not available.
+However if the `Task` being attempted has a default value for that parameter, it can still run,
+using the default when result is not available.
+
+### User Stories
+
+A `Pipeline` author can send `Pull Request` ID to finally task after building the source code and
+running unit/integration tests. The finally `Task` can send Slack notification with that ID and
+the result of the build.
```

Comment:
I wonder if it would make sense to replace this with a user story about clean up; in the motivation section we only discuss cleanup, and this particular user story has some other questions, since a finally task that is notifying might want to run even when the task it depends on fails (to update about the failure)

Reply:
Done 👍